core/jvm/src/main/java/zio/internal/Mailbox.java

```diff
@@ -0,0 +1,85 @@
+package zio.internal;
+
+import java.io.Serializable;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * An unbounded MPSC (multi-producers single consumer) queue that orders
+ * elements FIFO (first-in-first-out).
+ * 
+ * @apiNote The queue is thread-safe provided {@code poll} is invoked by the
+ *          single consumer (thread).
+ * 
+ * @implNote The implementation employs an algorithm described in <a href=
+ *           "https://www.1024cores.net/home/lock-free-algorithms/queues/non-intrusive-mpsc-node-based-queue">
+ *           Non-intrusive MPSC node-based queue</a> by D. Vyukov.
+ */
+final public class Mailbox<A> implements Serializable {
```

Comment:
let's make this private[zio] so we're not bound to keep this class or interface forever (future work may completely change the way FiberRuntime interacts with the mailbox)